Tx events subscription

viquezclaudio opened the pull request:
Ability to subscribe to some specific transaction addresses.
This functionality is mainly used by the web client, and should perform those subscription requests to full node and history nodes

viquezclaudio committed:
Create the RemoteEventDispatcher

This component is used to handle requests from other peers to subscribe to particular addresses

viquezclaudio committed:
Implement Remote Events for consensus

sisou commented on consensus/src/messages/handlers.rs:
```diff
@@ -10,16 +10,20 @@ use nimiq_block::Block;
-impl Handle<MacroChain, BlockchainProxy> for RequestMacroChain {
-    fn handle(&self, blockchain: &BlockchainProxy) -> MacroChain {
+impl<N: Network> Handle<N, MacroChain, BlockchainProxy> for RequestMacroChain {
+    fn handle(&self, _peer_id: N::PeerId, blockchain: &BlockchainProxy) -> MacroChain {
```
The `PeerId` is not used in any of the handlers. Everywhere the argument is prefixed with an underscore: `_peer_id`. Why did you add it, if it's not used anywhere?

viquezclaudio replied:
It's used by  `RequestSubscribeToAddress` to keep track of which addresses are being subscribed by specific peers.

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
         Ok(transactions)
     }
+
+    pub async fn subscribe_transaction_addresses(
```
```suggestion
    pub async fn subscribe_to_addresses(
```

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
+                RequestError::OutboundRequest(OutboundRequestError::SendError)
+            });
+        // Subscribe to any peer that could provide the necessary services
+        for peer_id in peers.unwrap() {
```
This subscribes to the current peers.

1. How do we subscribe at new peers that connect after this method call?
2. What happens when I call this method with the same addresses again? Do I get double notifications?

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
+                        crate::SubscribebyAdressErrors::Other => todo!(),
+                    },
+                },
+                Err(_) => {
+                    //Try with the next peer
+                    continue;
+                }
+            }
+        }
+    }
```
This method should return a result that signals if the subscription to at least one peer worked.

sisou commented on consensus/src/consensus/mod.rs:
```diff
@@ -106,8 +132,17 @@ impl<N: Network> Consensus<N> {
     ) -> Self {
         let (tx, _rx) = broadcast(256);
 
+        //Setup the remote events handler
```
You should add a linter script that finds these missing spaces :D 
```suggestion
        // Setup the remote events handler
```

sisou commented on consensus/src/consensus/mod.rs:
```diff
@@ -125,6 +160,24 @@ impl<N: Network> Consensus<N> {
+        // We spawn the Remote Event Dispatcher into its own task (this is only available for full nodes and history nodes)
+
+        match blockchain {
+            #[cfg(feature = "full")]
+            BlockchainProxy::Full(blockchain) => {
```
Do you need the `cfg` switch here? It's already matching against the blockchain type.

viquezclaudio replied:
Yes, because the "`BlockchainProxy::Full`" type is not defined in the web client.

sisou commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,312 @@
+        }
+        //2. Add the new ones
+        for address in new_addresses.iter() {
+            if let Some(peers) = self.subscriptions.get_mut(&address) {
```
Clippy warning here

sisou commented on consensus/src/error.rs:
```diff
@@ -15,3 +16,20 @@ pub enum SyncError {
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Error, Eq, PartialEq, Serialize, Deserialize)]
+pub enum SubscribebyAdressErrors {
```
```suggestion
pub enum SubscribeToAdressesError {
```

sisou commented on consensus/src/error.rs:
```diff
@@ -15,3 +16,20 @@ pub enum SyncError {
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Error, Eq, PartialEq, Serialize, Deserialize)]
+pub enum SubscribebyAdressErrors {
+    /// Already attending to many peers
+    #[error("TooManyPeers")]
+    TooManyPeers = 1,
+    /// Already attending to many peers
+    #[error("TooManyAddresses")]
+    TooManyAddresses = 2,
+    /// Some of the provided parameters/operations was invalid
+    #[error("InvalidOperation")]
+    InvalidOperation = 3,
+    /// Another type of error
+    #[error("Other")]
+    Other = 4,
```
```suggestion
    /// Already attending to too many peers
    #[error("Too many peers")]
    TooManyPeers = 1,
    /// Already attending to too many addresses
    #[error("Too many addresses")]
    TooManyAddresses = 2,
    /// Some of the provided parameters/operations are invalid
    #[error("Invalid operation")]
    InvalidOperation = 3,
    /// Another type of error not covered by the previous error types
    #[error("Other")]
    Other = 4,
```

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -248,3 +254,53 @@ pub struct ResponseTransactionReceiptsByAddress {
+pub enum TxAddressSubscriptionOperation {
+    /// Start the subscription, with the provided addresses
+    Initiate,
+    /// Update a current subscription, remplacing the existing one with the provided addresses
```
```suggestion
    /// Update a current subscription, replacing the existing one with the provided addresses
```

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -248,3 +254,53 @@ pub struct ResponseTransactionReceiptsByAddress {
+
+/// This request is used to initiate/update/terminate a subscription to events of some given addresses
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestSubscribeToAddress {
```
```suggestion
pub struct RequestSubscribeToAddresses {
```

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -248,3 +254,53 @@ pub struct ResponseTransactionReceiptsByAddress {
+/// Operations supported for the txn events by address subscription
+#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TxAddressSubscriptionOperation {
```
```suggestion
pub enum AddressSubscriptionOperation {
```

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -35,6 +37,10 @@ pub const MAX_REQUEST_RESPONSE_HEAD: u32 = 1000;
 /// The max number of Transactions proof requests per peer.
 pub const MAX_REQUEST_TRANSACTIONS_BY_ADDRESS: u32 = 1000;
+/// The max number of Subscribe to address requests per peer.
+pub const MAX_REQUEST_SUBSCRIBE_BY_ADDRESS: u32 = 10;
```
```suggestion
pub const MAX_REQUEST_SUBSCRIBE_TO_ADDRESSES: u32 = 10;
```

sisou replied:
Are these numbers (`10`) per a certain timeframe? Or in total?

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -35,6 +37,10 @@ pub const MAX_REQUEST_RESPONSE_HEAD: u32 = 1000;
+/// The max number of Subscribe to address requests per peer.
+pub const MAX_REQUEST_SUBSCRIBE_BY_ADDRESS: u32 = 10;
+/// The max number of Address notifications per peer.
+pub const MAX_ADDRESS_NOTIFICATION: u32 = 100;
```
```suggestion
pub const MAX_ADDRESS_NOTIFICATIONS: u32 = 100;
```

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
+        &self,
+        addresses: Vec<Address>,
+        min_peers: usize,
+        _max: Option<u16>,
```
Remove the `_max` argument?

sisou commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,312 @@
+                        result: Err(InvalidOperation),
+                    };
+                }
+                if self.addresses.len() > MAX_SUBSCRIBED_PEERS_ADDRESSES {
+                    return ResponseRequestTransactionsByAddress {
+                        result: Err(TooManyAddresses),
+                    };
+                }
```
This check in the `::Update` section fails to accept update requests that _reduce_ the number of subscribed addresses. That should be allowed, no?

sisou commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,312 @@
+                state.write().remove_peer(&peer_id);
+            }
+        }
+        ResponseRequestTransactionsByAddress { result: Ok(()) }
```
Usually subscriptions return a subscription ID that can then be used to manage that subscription. Did you decide against that pattern on purpose?

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
+            match response {
+                Ok(response) => match response.result {
+                    Ok(_) => {
+                        //Done, we are subscribed to a peer
```
```suggestion
                        // Done, we are subscribed at the peer
```

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
+                    },
+                },
+                Err(_) => {
+                    //Try with the next peer
```
```suggestion
                    // Try with the next peer
```

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -235,4 +246,57 @@ impl<N: Network> ConsensusProxy<N> {
+                .network
+                .request::<RequestSubscribeToAddress>(
+                    RequestSubscribeToAddress {
+                        operation: TxAddressSubscriptionOperation::Initiate,
```
Here the `::Initiate` type is used. The other types `::Update` and `::Terminate` are not used (yet). Do you have a plan to use those, too?

sisou reviewed (commented):
I cannot review the low-level integration and database access and stuff, but I tried to review the top-level API.

viquezclaudio committed:
Gossipsub subtopics

Create the concept of Gossipsub subtopics:
A subtopic is a topic with a subtopic name attached as a suffix to it.
For instance, they can be used by addresses subscriptions:
Address subscription is the main topic, and each specific peer_id is a new subtopic
where the interesting addresses, for that peer, are delivered

viquezclaudio committed:
Use a gossipsub subtopic for address subscriptions

Use a peer-specific gossipsub subtopic to handle interesting address subscriptions.
This creates the necessary logic to handle interesting addresses from blockchain extend events

jsdanielh commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -439,4 +455,99 @@ impl<N: Network> ConsensusProxy<N> {
+        min_peers: usize,
+        peer_id: Option<N::PeerId>,
+    ) -> Result<(), RequestError> {
+        //If we are provided a peer_id we perform the request only to this specific peer
```
```suggestion
        // If we are provided a peer_id we perform the request only to this specific peer
```

jsdanielh commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -439,4 +455,99 @@ impl<N: Network> ConsensusProxy<N> {
+            match response {
+                Ok(response) => match response.result {
+                    Ok(_) => {
+                        //Done, we are subscribed at least to one peer, continue with the next one
```
```suggestion
                        // Done, we are subscribed at least to one peer, continue with the next one
```

jsdanielh commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -439,4 +455,99 @@ impl<N: Network> ConsensusProxy<N> {
+        addresses: Vec<Address>,
+        min_peers: usize,
+    ) -> Result<(), RequestError> {
+        // Unsubscribe to all peers
```
```suggestion
        // Unsubscribe from all peers
```

jsdanielh commented on consensus/src/consensus/mod.rs:
```diff
@@ -7,6 +7,7 @@ use std::time::Duration;
 
 use futures::{FutureExt, StreamExt};
 use instant::Instant;
+use nimiq_hash::Blake2bHash;
```
This should be in the next block of `use`s

jsdanielh replied:
This should be in the next block of `use`s

jsdanielh commented on consensus/src/consensus/mod.rs:
```diff
@@ -128,6 +153,24 @@ impl<N: Network> Consensus<N> {
+                tokio::spawn(remote_event_dispatcher);
+            }
+            BlockchainProxy::Light(_) => {
+                //The light blockchain does not provide this functionality
```
```suggestion
                // The light blockchain does not provide this functionality
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use std::task::{Context, Poll};
+
+use nimiq_blockchain::Blockchain;
+use nimiq_blockchain_interface::AbstractBlockchain;
+use nimiq_blockchain_interface::BlockchainEvent;
```
Collapse them in a single `use`

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use nimiq_blockchain_interface::AbstractBlockchain;
+use nimiq_blockchain_interface::BlockchainEvent;
+use nimiq_keys::Address;
+use nimiq_network_interface::request::Handle;
+use nimiq_network_interface::{network::Network, request::request_handler};
```
Collapse them in a single `use`

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use nimiq_hash::{Blake2bHash, Hash};
+use nimiq_network_interface::network::{NetworkEvent, SubscribeEvents, Topic};
+use parking_lot::RwLock;
+use std::collections::{HashMap, HashSet};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
```
Collapse these `use`s and get it to the beginning of the file

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use beserial::{Deserialize, Serialize};
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use nimiq_hash::{Blake2bHash, Hash};
```
Move this `use` to the block of nimiq `use`s

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use beserial::{Deserialize, Serialize};
+use futures::stream::BoxStream;
+use futures::StreamExt;
```
Collapse these two `use`s

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use beserial::{Deserialize, Serialize};
```
Move this `use` to the block of nimiq `use`s

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use nimiq_hash::{Blake2bHash, Hash};
+use nimiq_network_interface::network::{NetworkEvent, SubscribeEvents, Topic};
```
Move this `use` to the block of nimiq `use`s

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+use crate::messages::{AddressSubscriptionOperation, ResponseSubscribeToAddress};
+use crate::SubscribeToAdressesError::*;
+
+//The max number of peers that can be subscribed.
```
```suggestion
/// The max number of peers that can be subscribed.
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+
+//The max number of peers that can be subscribed.
+pub const MAX_SUBSCRIBED_PEERS: usize = 5;
+//The max number of addresses that can be subscribed, per peer
```
```suggestion
/// The max number of addresses that can be subscribed, per peer
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+
+    const BUFFER_SIZE: usize = 1024;
+    const NAME: &'static str = "transactions";
+    const VALIDATE: bool = true;
```
Since we are not doing any validation, this should be `false`

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+    type Item = AddressNotification;
+
+    const BUFFER_SIZE: usize = 1024;
+    const NAME: &'static str = "transactions";
```
Maybe rename it to avoid possible conflicts with the regular `transaction` topic

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+            }
+
+            AddressSubscriptionOperation::Unsubscribe => {
+                //If we don't know this peer, we don't do anything
```
```suggestion
                // If we don't know this peer, we don't do anything
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+        ResponseSubscribeToAddress { result: Ok(()) }
+    }
+}
+
```
Missing `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,329 @@
+}
+
+pub struct RemoteEventDispatcherState<N: Network> {
+    // HashMap containing a mapping from peers to their interesting addresses
```
```suggestion
    /// HashMap containing a mapping from peers to their interesting addresses
```

viquezclaudio committed:
Cleanup and PR comments

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+            }
+
+            AddressSubscriptionOperation::Unsubscribe => {
+                //If we don't know this peer, we don't do anything
```
```suggestion
                // If we don't know this peer, we don't do anything
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+/// The state that is mantained by the remote event dispatcher:
+/// essentially the addresses and peers that are subscribed to us.
+pub struct RemoteEventDispatcherState<N: Network> {
+    // HashMap containing a mapping from peers to their interesting addresses
```
```suggestion
    /// HashMap containing a mapping from peers to their interesting addresses
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+            subscriptions: HashMap::new(),
+        }
+    }
+
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+    pub fn contains_peer(&self, peer_id: &N::PeerId) -> bool {
+        self.subscribed_peers.contains_key(peer_id)
+    }
+
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+    pub fn number_of_peers(&self) -> usize {
+        self.subscribed_peers.len()
+    }
+
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+            }
+        }
+    }
+
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+        // Finally remove the peer
+        self.subscribed_peers.remove(peer_id);
+    }
+
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+            }
+        }
+    }
+
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+/// because they are interested in different events.
+/// This is mainly used by light clients who want to fetch information from Full & History nodes
+pub struct RemoteEventDispatcher<N: Network> {
+    /// The state that we mantain, with the peers and their interesting addresses
```
```suggestion
    /// The state that we maintain, with the peers and their interesting addresses
```

jsdanielh commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,300 @@
+/// This is mainly used by light clients who want to fetch information from Full & History nodes
+pub struct RemoteEventDispatcher<N: Network> {
+    /// The state that we mantain, with the peers and their interesting addresses
+    pub state: Arc<RwLock<RemoteEventDispatcherState<N>>>,
```
Do we need it to be `pub`?

jsdanielh commented on consensus/src/error.rs:
```diff
@@ -1,3 +1,4 @@
+use beserial::{Deserialize, Serialize};
```
Move this `use` to the nimiq `use` block

jsdanielh commented on consensus/src/error.rs:
```diff
@@ -15,3 +16,20 @@ pub enum SyncError {
 }
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Error, Eq, PartialEq, Serialize, Deserialize)]
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/error.rs:
```diff
@@ -15,3 +16,20 @@ pub enum SyncError {
+#[derive(Clone, Copy, Debug, Error, Eq, PartialEq, Serialize, Deserialize)]
+pub enum SubscribeToAdressesError {
+    /// Already attending too many peers
+    #[error("TooManyPeers")]
```
```suggestion
    #[error("Too many peers")]
```

jsdanielh commented on consensus/src/error.rs:
```diff
@@ -15,3 +16,20 @@ pub enum SyncError {
+    /// Already attending too many peers
+    #[error("TooManyPeers")]
+    TooManyPeers = 1,
+    /// Already attending too many peers
```
The same as above?

jsdanielh commented on consensus/src/error.rs:
```diff
@@ -15,3 +16,20 @@ pub enum SyncError {
+    #[error("TooManyPeers")]
+    TooManyPeers = 1,
+    /// Already attending too many peers
+    #[error("TooManyAddresses")]
```
```suggestion
    #[error("Too many addresses")]
```

jsdanielh commented on consensus/src/lib.rs:
```diff
@@ -2,8 +2,9 @@
-pub use consensus::{Consensus, ConsensusEvent};
+pub use consensus::{Consensus, ConsensusEvent, RemoteEvent};
 pub use error::Error;
+pub use error::SubscribeToAdressesError;
```
Collapse it in the previous `use`

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -294,3 +303,67 @@ impl RequestCommon for RequestBlocksProof {
+    Unsubscribe,
+}
+
+/// This request is used to subscribe or unsuscribe from specific addresses.
```
```suggestion
/// This request is used to subscribe or unsubscribe from specific addresses.
```

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -294,3 +303,67 @@ impl RequestCommon for RequestBlocksProof {
+
+impl RequestCommon for RequestSubscribeToAddress {
+    type Kind = RequestMarker;
+    const TYPE_ID: u16 = 215;
```
This is already in use

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -294,3 +303,67 @@ impl RequestCommon for RequestBlocksProof {
+    const MAX_REQUESTS: u32 = MAX_REQUEST_SUBSCRIBE_BY_ADDRESS;
+}
+
+#[derive(Serialize, Deserialize)]
```
Add `rustdoc` documentation

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -294,3 +303,67 @@ impl RequestCommon for RequestBlocksProof {
+#[derive(Serialize, Deserialize, Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum NotificationEvent {
+    BlockchainExtend,
```
Add member `rustdoc` documentation

jsdanielh commented on consensus/tests/consensus_proxy.rs:
```diff
@@ -17,11 +17,10 @@ use nimiq_test_utils::{
 use nimiq_zkp_component::ZKPComponent;
 use parking_lot::{Mutex, RwLock};
-use std::any::Any;
 use std::str::FromStr;
 use std::sync::Arc;
```
We should move the `std` `use`s to the beginning and separate the one for `parking_lot`

jsdanielh commented on consensus/tests/state_live_sync.rs:
```diff
@@ -605,8 +605,12 @@ async fn can_reset_chain_of_chunks() {
-            mock_node.set_chunk_handler(Some(|request, blockchain| {
-                let mut chunk = request.handle(blockchain);
+            mock_node.set_chunk_handler(Some(|peer_id, request, blockchain| {
+                let mut chunk = <RequestChunk as Handle<
```
Do we really need the typed `handle`?

viquezclaudio replied:
Yes, type annotations are needed in this case

jsdanielh commented on consensus/tests/state_live_sync.rs:
```diff
@@ -687,8 +691,12 @@ async fn can_remove_chunks_related_to_invalid_blocks() {
-        let mut response = request.handle(blockchain_proxy);
+    mock_node.set_missing_block_handler(Some(|mock_id, request, blockchain_proxy| {
+        let mut response = <RequestMissingBlocks as Handle<
+            MockNetwork,
```
Do we really need the typed `handle`?

jsdanielh commented on consensus/tests/state_live_sync.rs:
```diff
@@ -785,8 +793,12 @@ async fn clears_buffer_after_macro_block() {
-        let mut chunk = request.handle(blockchain);
+    mock_node.set_chunk_handler(Some(|mock_id, request, blockchain| {
+        let mut chunk = <RequestChunk as nimiq_network_interface::request::Handle<
+            MockNetwork,
```
Do we really need the typed `handle`?

jsdanielh commented on network-interface/src/request/mod.rs:
```diff
@@ -243,3 +249,34 @@ pub fn request_handler<
+                    tokio::spawn(async move {
+                        log::trace!("{:?} {:#?}", peer_id, msg);
+
+                        //Messages do not have a response (so the response is ignored)
```
```suggestion
                        // Messages do not have a response (so the response is ignored)
```

jsdanielh commented on network-interface/src/request/mod.rs:
```diff
@@ -243,3 +249,34 @@ pub fn request_handler<
             .await
     }
 }
+
```
Add `rustdoc` documentation

jsdanielh commented on network-mock/src/network.rs:
```diff
@@ -455,7 +455,126 @@ impl Network for MockNetwork {
 
         if self.is_connected.load(Ordering::SeqCst) {
-            if let Some(topic) = hub.get_topic(topic_name) {
+            if let Some(topic) = hub.get_topic(&topic_name) {
```
I'd have expected to re-use some of the already existing functionality to not duplicate code

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -439,4 +455,101 @@ impl<N: Network> ConsensusProxy<N> {
+        addresses: Vec<Address>,
+        min_peers: usize,
+    ) -> Result<(), RequestError> {
+        // Unsubscribe addresses from all peers
+        // Note: this does not mean that we will fully unsuscribe from a peer,
+        // we will unsuscribe only from the addresses that were suplied to this function
```
```suggestion
        // Unsubscribe given addresses from all peers
        // Note: this does not mean that we will fully unsubscribe from a peer,
        // we will unsubscribe only from the addresses that were supplied to this function
```

sisou commented on consensus/src/consensus/consensus_proxy.rs:
```diff
@@ -439,4 +455,101 @@ impl<N: Network> ConsensusProxy<N> {
+                )
+                .await;
+
+            // We don't care about the response, we just unsuscribe addresses from peers
```
```suggestion
            // We don't care about the response, we just unsubscribe addresses from peers
```

viquezclaudio committed:
Refactor some network functions to avoid duplication

viquezclaudio committed:
Notify rebranch, finalized and epoch finalized events

Fixed some clippy warnings as well

sisou commented on consensus/src/consensus/remote_event_dispatcher.rs:
```diff
@@ -0,0 +1,315 @@
+    pub fn remove_peer(&mut self, peer_id: &N::PeerId) {
+        if let Some(peer_addresses) = self.subscribed_peers.get(peer_id) {
+            // Obtain the addresses that are interested to this peer and remove the peer from those addresses.
+            peer_addresses.into_iter().for_each(|address| {
```
Clippy warning here: this `.into_iter()` call is equivalent to `.iter()` and will not consume the `HashSet`

sisou reviewed (commented):
6 clippy warnings that "`to_string` applied to a type that implements `Display` in `format!` args".

jsdanielh reviewed (changes requested):
There are some clippy errors
